src/main/java/com/playmonumenta/networkrelay/RemotePlayerManagerBungee.java

```diff
@@ -0,0 +1,216 @@
+import org.jetbrains.annotations.Nullable;
+
+public final class RemotePlayerManagerBungee extends RemotePlayerManagerAbstraction implements Listener {
+	private static @MonotonicNonNull RemotePlayerManagerBungee INSTANCE = null;
+
+	private RemotePlayerManagerBungee() {
+		INSTANCE = this;
+		try {
+			for (String shard : NetworkRelayAPI.getOnlineShardNames()) {
+				MMLog.fine(() -> "Registering shard " + shard);
+				registerServer(shard);
+			}
+		} catch (Exception ex) {
+			MMLog.severe(() -> "Failed to get remote shard names");
+			throw new RuntimeException("Failed to get remote shard names", ex);
+		}
+
+		try {
+			NetworkRelayAPI.sendExpiringBroadcastMessage(REMOTE_PLAYER_REFRESH_CHANNEL,
+				new JsonObject(),
+				REMOTE_PLAYER_MESSAGE_TTL);
+		} catch (Exception ex) {
+			MMLog.severe(() -> "Failed to broadcast to channel " + REMOTE_PLAYER_REFRESH_CHANNEL);
+		}
+	}
+
+	static RemotePlayerManagerBungee getInstance() {
+		if (INSTANCE == null) {
+			INSTANCE = new RemotePlayerManagerBungee();
+		}
+		return INSTANCE;
+	}
+
+	static String getServerId() {
+		@Nullable String shardName = null;
+		try {
+			shardName = NetworkRelayAPI.getShardName();
+		} catch (Exception e) {
+			MMLog.severe(() -> "Failed to get shard name");
+		}
+		if (shardName == null) {
+			throw new RuntimeException("Got null shard name");
+		}
+		return shardName;
+	}
+
+	static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline) {
+		@Nullable String targetShard = player.getServer() != null ? player.getServer().getInfo().getName() : ""; // TODO: add a way to get the shard name. IT IS MUCH EASIER IN VELOCITY - usb
+		return fromLocal(player, isOnline, targetShard);
+	}
+
+	static RemotePlayerBungee fromLocal(ProxiedPlayer player, boolean isOnline, String targetShard) {
+		return new RemotePlayerBungee(
+			getServerId(),
+			player.getUniqueId(),
+			player.getName(),
+			isOnline,
+			null,
+			targetShard
+		);
+	}
+
+	void refreshLocalPlayers() {
+		for (ProxiedPlayer player : ProxyServer.getInstance().getPlayers()) {
+			refreshLocalPlayer(player);
+		}
+	}
+
+	// Run this on local players whenever their information is out of date
+	void refreshLocalPlayer(ProxiedPlayer player) {
+		MMLog.fine(() -> "Refreshing local player " + player.getName());
+		RemotePlayerBungee localPlayer = fromLocal(player, true);
+
+		// update local player with data
+		if (updatePlayerLocal(localPlayer, false)) {
+			localPlayer.broadcast();
+		}
+	}
+
+	void remotePlayerChange(JsonObject data) {
+		if (data == null) {
+			MMLog.severe(() -> "Null player data recieved from an unknown source!");
+			return;
+		}
+		RemotePlayerAbstraction player = RemotePlayerAbstraction.from(data);
+		if (player == null) {
+			// something really bad happened
+			MMLog.severe(() -> "Invalid player data recieved from an unknown source!");
+			return;
+		}
+
+		updatePlayerLocal(player, true);
+	}
+
+	boolean updatePlayerLocal(RemotePlayerAbstraction player, boolean isRemote) {
```

Comment:
This could use a comment block explaining the return value; it appears that it returns false if the local value is correct and doesn't need updating. Also consider the name `updateLocalCache` or similar, which should avoid confusion with why updating something local needs a flag for remote data. (Though that might be on me for how I named things the first time around.)